lib/ZUPnP.c

```diff
@@ -0,0 +1,85 @@
+#include <internal.h>
+
+#ifdef HAVE_UPNP
+#include <arpa/inet.h>
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/upnpcommands.h>
+
+static struct UPNPUrls __UPnP_urls = {.controlURL = NULL};
+static struct IGDdatas __UPnP_data;
+static int __UPnP_attempted = 0;
+static char* __UPnP_name = "Zephyr Client";
+
+void Z_InitUPnP_ZHM() {
+  struct UPNPDev * devlist;
+  int upnperror = 0;
+  if (__UPnP_active) {
+    // tried to initialize twice
+    return;
+  }
+  devlist = upnpDiscover(
+			 2000,
+			 NULL/*multicast interface*/,
+			 NULL/*minissdpd socket path*/,
+			 UPNP_LOCAL_PORT_ANY/*sameport*/,
+			 0/*ipv6*/,
+			 2/*TTL*/,
+			 &upnperror);
+  if (devlist) {
+    int igdfound = UPNP_GetValidIGD(devlist, &__UPnP_urls, &__UPnP_data, NULL, 0);
+    if (igdfound) {
+      __UPnP_rooturl = __UPnP_urls.rootdescURL;
+      char extIpAddr[16];
+      if (UPNP_GetExternalIPAddress(__UPnP_urls.controlURL, __UPnP_data.first.servicetype, extIpAddr) == 0) {
+	struct in_addr ext_addr;
+	if (inet_aton(extIpAddr, &ext_addr)) {
+	  __My_addr = ext_addr;
+	  __UPnP_active = 1;
+	}
+      }
+    }
+    freeUPNPDevlist(devlist);
+  }
+  __UPnP_name = "Zephyr Host Manager";
+  Z_InitUPnP();
+}
+
+void Z_InitUPnP() {
+  if (__UPnP_attempted) {
+    return;
+  }
+  __UPnP_attempted = 1;
+  if (__UPnP_rooturl && !__UPnP_active) {
+    __UPnP_active = UPNP_GetIGDFromUrl(__UPnP_rooturl, &__UPnP_urls, &__UPnP_data, NULL, 0);
+  }
+  if (__UPnP_active) {
+    char port_str[16];
```

Comment:
Why 16? Port are only 0..65535, so shouldn't [8] be sufficient?

Reply:
This is copied verbatim from the library.